### PR TITLE
Addition of Efficient/Lagrangian Point-Mass Filter 

### PR DESCRIPTION
**Pull request by pesslovany**
J. Matoušek, J. Duník and M. Brandner, "Design of Efficient Point-Mass Filter with Terrain Aided Navigation Illustration," 2023 26th International Conference on Information Fusion (FUSION), Charleston, SC, USA, 2023, pp. 1-8, doi: 10.23919/FUSION52260.2023.10224172. 

J. Matoušek, J. Duník, M. Brandner, Chan Gook Park, Yeongkwon Choe,
Efficient Point Mass Predictor for Continuous and Discrete Models with Linear Dynamics,IFAC-PapersOnLine,Volume 56, Issue 2,2023,Pages 5937-5942,ISSN 2405-8963,https://doi.org/10.1016/j.ifacol.2023.10.621.

J. Matoušek, J. Duník, O. Straka,
Lagrangian Grid-Based Filters with Application to Terrain-Aided Navigation
IEEE Signal Processing Magazine - Accepted, to be published

Added points mass updater, point mass predictor, and point mass state.

### CODE REVIEW

File: stonesoup/types/prediction.py [hunk 1]
```diff
@@ -142,13 +142,27 @@ class ParticleStatePrediction(Prediction, ParticleState):
     """
 
 
+class PointMassStatePrediction(Prediction, PointMassState):
+    """ParticleStatePrediction type
+
+    This is a simple Particle state prediction object.
+    """
```
Comment (hpritchett-dstl) [inline stonesoup/types/prediction.py:149]: This docstring is for ParticleStatePrediction, could you make a minor change to make it relevant for PointMassStatePrediction ?

File: functions.py
[the remark this answers is not in the document]
Comment (pesslovany) [reply functions.py:1]: I have deleted the file.

File: stonesoup/updater/pointMass.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+    Perform an update by multiplying grid points weights by PDF of measurement
+    model
+    """
+    sFactor: float = Property(default=4, doc="How many sigma to cover by the grid")
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    # @profile
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/updater/pointMass.py:21]: Deleted

File: stonesoup/updater/pointMass.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+from .base import Updater
+
+
+class PointMassUpdater(Updater):
+    """Point mass Updater
+
+    Perform an update by multiplying grid points weights by PDF of measurement
+    model
+    """
+    sFactor: float = Property(default=4, doc="How many sigma to cover by the grid")
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/updater/pointMass.py:16]: Done

File: stonesoup/types/state.py [hunk 4]
```diff
@@ -158,31 +168,81 @@ def from_state(
...
+    def ndim(self):
+        """The number of dimensions represented by the state."""
+        return self.state_vector.shape[0]
+
+    @clearable_cached_property("state_vector")
+    def mean(self):
+        """Sample mean for particles"""
+        return np.hstack(self.state_vector @ self.weight * np.prod(self.grid_delta))
+
+    # @profile
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/types/state.py:218]: Deleted

File: stonesoup/types/prediction.py (hunk 1 above)
Comment (pesslovany) [reply stonesoup/types/prediction.py:149]: I am also lost what at what is the purpose of the prediction object, but everything seems to be working....

File: stonesoup/predictor/pointMass.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+from ..types.array import StateVectors
+from .base import Predictor
+
+pio.renderers.default = "browser"
+
+
+class PointMassPredictor(Predictor):
+    """ParticlePredictor class
+
+    An implementation of a Particle Filter predictor.
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/predictor/pointMass.py:20]: Fixed. Sorry. As said below. :)

File: stonesoup/predictor/pointMass.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    """
+
+    sFactor: float = Property(default=4, doc="How many sigma to cover by the grid")
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/predictor/pointMass.py:23]: Fixed

File: stonesoup/functions/__init__.py [hunk 7]
```diff
@@ -1,14 +1,83 @@
...
+    gridStep = []  # Reset gridStep for each cycle
+    for ind3 in range(0, nx):  # Creation of propagated grid
+        # New grid with middle in 0
+        gridDim.append(np.linspace(-gridBound[ind3], gridBound[ind3], Npa[ind3]))
+        gridStep.append(np.absolute(gridDim[ind3][0] - gridDim[ind3][1]))  # Grid step
+
+    combvec_predGrid = np.array(list(itertools.product(*gridDim)))
+    predGrid_pom = np.dot(eigVect, combvec_predGrid.T)
+    size_pom = np.size(predGrid_pom, 1)
+    # Grid rotation by eigenvectors and traslation to the counted unscented mean
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/functions/__init__.py:75]: Fixed

File: stonesoup/functions/__init__.py [hunk 8]
```diff
@@ -1,14 +1,83 @@
...
+    gridDim : list of numpy.ndarrays
+        grid coordinates per dimension before rotation and translation 
+    xp_aux : numpy.ndarray
+        grid center
+    eigVect : numpy.ndarray
+        eigenvectors describing the rotation of the grid
+        
+    """
+    gridDim = np.zeros((nx, Npa[0]))
+    gridStep = np.zeros(nx)
```
[the remark this answers is not in the document]
Comment (pesslovany) [reply stonesoup/functions/__init__.py:49]: Done

**Comment by pesslovany** [comment]
Thank you for your comments. I think I have fixed the codes based on them. I have also merged my branch with the latest changes. And it seems I have fixed all the issues so the code should run all of the testes successfully. 

Hopefully we can finish the pull request. :)

File: stonesoup/predictor/pointMass.py [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    def predict(self, prior, timestamp=None, **kwargs):
+        """Point Mass Filter prediction step
+
+        Parameters
+        ----------
+        prior : :class:`~.Point mass state`
+            A prior state object
+        timestamp: :class:`datetime.datetime`, optional
+            A timestamp signifying when the prediction is performed
+            (the default is `1`)
```
Comment (hpritchett-dstl) [inline stonesoup/predictor/pointMass.py:34]:
```suggestion
```

File: stonesoup/updater/pointMass.py
Comment (hpritchett-dstl) [inline stonesoup/updater/pointMass.py:1]: to maintain python file naming convention, could you change the name of this file to pointmass.py? Note that the import for this in test_pointmass.py will then need to be changed to reflect this :)

File: stonesoup/predictor/pointMass.py
Comment (hpritchett-dstl) [inline stonesoup/predictor/pointMass.py:1]: to maintain python file naming convention, could you change the name of this file to pointmass.py? Note that the import for this in test_pointmass.py.

**Review by hpritchett-dstl — COMMENTED** [review]
Some final comments - just need to change the name of the predictor and updater files to maintain python naming conventions.

Once that's done I'm happy to approve this. This is some great work!

File: stonesoup/predictor/pointMass.py (hunk 9 above)
Comment (pesslovany) [reply stonesoup/predictor/pointMass.py:34]: Done

File: stonesoup/updater/pointMass.py
Comment (pesslovany) [reply stonesoup/updater/pointMass.py:1]: Done. Thank you for mentioning the test.

File: stonesoup/predictor/pointMass.py
Comment (pesslovany) [reply stonesoup/predictor/pointMass.py:1]: Done

**Comment by pesslovany** [comment]
Comments should be addressed. :)

**Review by hpritchett-dstl — APPROVED** [review]

**Pull request closed without merging**